Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,728 @@
+More on this later.
+
+The bottom section contains some boilerplate that Alex requires us to write. These include a data type that needs to be called `AlexUserState`, a value with the initial state called `alexInitUserState`, and a value called `alexEOF`, which instructs Alex how to build the EOF (End-Of-File) token, reached when Alex has finished lexing the input string.
+
+We add some additional datatypes: `Range`, `Token`, and `TokenType`, which we will use throughout the article to describe the tokens that we've successfully created, as well as their positions. Saving the ranges is unnecessary, but it's an excellent addition when reporting errors.
+
+Finally, for the EOF token, we use the `alexGetInput` action to retrieve the current position of the scanner and provide it to the token.
+
+We should provide one token type for each lexeme in the grammar.
+You are free to choose the available tokens. Still, for this tutorial, we will have the following: operators, keywords, literals (strings and integers), identifiers, a token representing the end-of-file (EOF), etc.
+
+Our token types will be as such:
+
+```haskell
+data TokenType
+  -- Identifiers
+  = Identifier ByteString
+  -- Constants
+  | String ByteString
+  | Integer Integer
+  -- Keywords
+  | Let
+  | In
+  | If
+  | Then
+  | Else
+  -- Arithmetic operators
+  | Plus
+  | Minus
+  | Times
+  | Divide
+  -- Comparison operators
+  | Eq
+  | Neq
+  | Lt
+  | Le
+  | Gt
+  | Ge
+  -- Logical operators
+  | And
+  | Or
+  -- Parenthesis
+  | LPar
+  | RPar
+  -- Lists
+  | Comma
+  | LBrack
+  | RBrack
+  -- Types
+  | Colon
+  | Arrow
+  -- EOF
+  | EOF
+  deriving (Eq, Show)
+```
+
+At the top of the file, make sure to include the appropriate imports:
+
+```haskell
+import Data.ByteString.Lazy.Char8 (ByteString)
+import qualified Data.ByteString.Lazy.Char8 as BS
+```
+
+### Regex and regex macros
+
+Alex uses regular expressions to match lexemes, so we will use them to recognize the various lexemes that should be available in our grammar.
+Let's start with a simple one: identifiers.
+We are free to choose the specification for identifiers, but let's use the following:
+
+> An identifier is a sequence of alphanumeric characters, primes ('), question marks (?), and underscores (_).
+> The identifier must begin with a letter or an underscore.
+
+Alex allows the creation of _character set macros_, which are shortcuts that you can use to avoid duplicating macros. We use `$NAME = CHARACTER_SET` to define a character set macro.
+
+```alex
+%wrapper "monadUserState-bytestring"
+
+$digit = [0-9]
+$alpha = [a-zA-Z]
+
+tokens :-
+```
+
+We may also combine multiple regex macros into one, which will be appropriately expanded in its equivalent regular expression.
+For example, here's how we can match an identifier:
+
+```alex
+$digit = [0-9]
+$alpha = [a-zA-Z]
+
+@id = ($alpha | \_) ($alpha | $digit | \_ | ' | \?)*
```

Comment:
This single quote could use a backslash, if only because common syntax highlighting grammars break on it otherwise:

```
@id = ($alpha | \_) ($alpha | $digit | \_ | \' | \?)*
```

Reply:
Done.